azure-functions/src/bindings/durable_orchestration_context.rs

```diff
@@ -24,210 +212,127 @@ use std::task::{Context, Poll};
+    pub fn current_time(&self) -> DateTime<Utc> {
+        let state = self.state.borrow();
+        state.history[state.orchestration_index].timestamp
     }
 
     #[doc(hidden)]
     pub fn execution_result(&self) -> Rc<RefCell<ExecutionResult>> {
-        self.result.clone()
+        self.state.borrow().result.clone()
     }
 
-    #[doc(hidden)]
+    /// Creates a future which represents a collection of the outputs of the futures given.
+    ///
+    /// The returned future will drive execution for all of its underlying futures,
+    /// collecting the results into a destination `Vec<T>` in the same order as they
+    /// were provided.
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
+    pub fn join_all<I>(&self, iter: I) -> JoinAll<I::Item>
+    where
+        I: IntoIterator,
+        I::Item: OrchestrationFuture,
+    {
+        JoinAll::new(self.state.clone(), iter)
+    }
+
+    /// Schedules an activity function for execution.
+    #[must_use = "futures do nothing unless you `.await` or poll them"]
     pub fn call_activity<D>(
         &mut self,
         activity_name: &str,
         data: D,
-    ) -> impl Future<Output = Result<Value, String>>
+    ) -> impl Future<Output = Result<Value, String>> + OrchestrationFuture
     where
         D: Into<Value>,
     {
-        self.result.borrow_mut().add_action(Action::CallActivity {
+        let mut state = self.state.borrow_mut();
+
+        // Push the action on the execution result
+        state.result.borrow_mut().push_action(Action::CallActivity {
             function_name: activity_name.to_string(),
             input: data.into(),
         });
 
-        let mut event_result = Option::<Result<Value, String>>::None; //ActionFuture(None); //Option::<ActionFuture<Result<Value, String>>>>::None;
+        let mut result: Option<Result<Value, String>> = None;
 
-        if let Some(scheduled) = self.find_scheduled_activity(activity_name) {
-            scheduled.processed = true;
+        // Attempt to resolve the activity
+        if let Some(scheduled) = state.find_scheduled_activity(activity_name) {
+            scheduled.is_processed = true;
 
             let id = scheduled.event_id;
-            if let Some(completed) = self.find_completed_activity(id) {
-                completed.processed = true;
-
-                let value = completed.result.clone().unwrap_or_default();
-                event_result = Some(Ok(value));
-            } else if let Some(failed) = self.find_failed_activity(id) {
-                failed.processed = true;
-
-                let message = failed.reason.clone().unwrap_or_default();
-                event_result = Some(Err(message));
+            if let Some(completed) = state.find_completed_activity(id) {
+                completed.is_processed = true;
+                result = Some(Ok(completed
+                    .result
+                    .as_ref()
+                    .map(|s| from_str(&s).unwrap_or_default())
```

Comment:
Any specific reason to parse it in place instead of when parsed during creation?

Reply:
It's always passed as a string in the data, rather than a JSON value. Likewise, the JavaScript implementation also parses it when it resolves the event.